db/import.go

```diff
@@ -263,29 +259,41 @@ func (db *Database) importDoc(docid string, body Body, isDelete bool, existingDo
 			bodyWithoutSpecialProps, wasStripped = stripSpecialProperties(body)
 			rawBodyForRevID, err = base.JSONMarshalCanonical(bodyWithoutSpecialProps)
 			if err != nil {
-				return nil, nil, nil, err
+				return nil, nil, false, nil, err
 			}
 		}
 
-		newRev = CreateRevIDWithBytes(generation, parentRev, rawBodyForRevID)
-		if err != nil {
-			return nil, nil, updatedExpiry, err
-		}
-		base.DebugfCtx(db.Ctx, base.KeyImport, "Created new rev ID for doc %q / %q", base.UD(newDoc.ID), newRev)
-		// body[BodyRev] = newRev
-		newDoc.RevID = newRev
-		err := doc.History.addRevision(newDoc.ID, RevInfo{ID: newRev, Parent: parentRev, Deleted: isDelete})
-		if err != nil {
-			base.InfofCtx(db.Ctx, base.KeyImport, "Error adding new rev ID for doc %q / %q, Error: %v", base.UD(newDoc.ID), newRev, err)
-		}
+		avoidRevIDGeneration = !doc.BodyChanged(existingDoc.Body)
```

Comment:
Can 'bodyChanged' be returned by doc.IsSGWrite, to avoid computing the crc32c of the existing body twice?

Reply:
I'm not sure this would improve a lot. Its not quite as simple as we're always computing the hash inside of IsSGWrite because in the case of an xattr change we currently only compute the body hash once using doc.BodyChanged and won't check the body at all inside of IsSGWrite. If we were to return BodyChanged from IsSGWrite we would need to always compute the body hash regardless of what happened to the xattr meaning every time we run IsSGWrite we're going to run this hash check. IsSGWrite gets run a few times for an import operation. I think the question would be in the case of non-xattr imports do we want to pay the cost of an additional hash or in the case of an xattr import do we want to pay the cost of multiple body hash checks most of which are 'wasted'.